File: runtimes/ios/WebToApp/OriginPolicy.swift
import Foundation

struct OriginPolicy {
    let allowedOrigins: Set<String>

    init(origins: [String], verifiedDomains: Set<String>) throws {
        guard !origins.isEmpty, origins.allSatisfy({ value in
            guard let url = URL(string: value), let host = url.host else { return false }
            return Self.origin(url) == value && verifiedDomains.contains(host)
        }) else { throw URLError(.unsupportedURL) }
        allowedOrigins = Set(origins)
    }

    func allows(_ url: URL) -> Bool {
        guard let origin = Self.origin(url) else { return false }
        return allowedOrigins.contains(origin)
    }

    static func origin(_ url: URL) -> String? {
        guard url.scheme == "https", url.user == nil, url.password == nil,
              let host = url.host, host == host.lowercased(),
              url.port == nil || url.port == 443 else { return nil }
        return "https://" + host
    }
}

File: runtimes/android/app/src/main/java/dev/webtoapp/runtime/OriginPolicy.kt
package dev.webtoapp.runtime

import java.net.URI

class OriginPolicy(origins: List<String>, verifiedDomains: Set<String>) {
    val allowedOrigins: Set<String> = origins.toSet()
    init {
        require(allowedOrigins.isNotEmpty())
        require(allowedOrigins.all {
            val uri = URI(it)
            origin(it) == it && uri.host in verifiedDomains
        }) { "Only exact verified HTTPS origins are accepted" }
    }

    fun allows(url: String): Boolean = origin(url) in allowedOrigins

    companion object {
        fun origin(value: String): String? = try {
            val uri = URI(value)
            if (uri.scheme != "https" || uri.rawUserInfo != null || uri.host == null ||
                uri.host != uri.host.lowercase() || uri.port !in listOf(-1, 443)) null
            else "https://" + uri.host
        } catch (_: Exception) { null }
    }
}
